File: unittests/ym/common/rng/testsuite.cpp
/**
 * @file    testsuite.cpp
 * @version 1.0.0
 * @author  Forrest Jablonski
 */

#include "ymdefs.h"
#include "testsuite.h"

#include "rng.h"
#include "textlogger.h"
#include "ymerror.h"

#include <algorithm>
#include <bitset>
#include <utility>
#include <vector>

/** TestSuite
 *
 * @brief Constructor.
 */
ym::unit::TestSuite::TestSuite(void)
   : TestSuiteBase("Rng")
{
   addTestCase<ZerosAndOnes>();
   addTestCase<UniformBins >();
}

/** run
 *
 * @brief Runs zeros and ones test.
 * 
 * @note Amount of zeros and ones should be about equal.
 * 
 * @note That's what a relationship is. We average our misery.
 *
 * @returns DataShuttle -- Important values acquired during run of test case.
 */
auto ym::unit::TestSuite::ZerosAndOnes::run([[maybe_unused]] DataShuttle const & InData) -> DataShuttle
{
   auto const SE = ymLogPushEnable(VG::UnitTest_Rng);

   ym::Prng rand;

   std::vector<std::pair<uint64, uint64>> setBitVector;

   for (auto i = 8_u64; i <= 24_u64; ++i)
   {
      auto const NIters   = (1_u64 << i);
      auto       nSetBits = 0_u64;
      for (auto j = 0_u64; j < NIters; ++j)
      {
         nSetBits += std::bitset<64>(rand.gen<uint64>()).count();
      }
      auto const NTrials = NIters * 64_u64;
      setBitVector.emplace_back(NTrials, std::min(nSetBits, NTrials - nSetBits));
   }

   return {
      {"SetBitVector", setBitVector}
   };
}

/** run
 *
 * @brief Runs uniform bins test.
 * 
 * @note Amount of entries in each bin should be about equal.
 *
 * @returns DataShuttle -- Important values acquired during run of test case.
 */
auto ym::unit::TestSuite::UniformBins::run([[maybe_unused]] DataShuttle const & InData) -> DataShuttle
{
   auto const SE = ymLogPushEnable(VG::UnitTest_Rng);

   ym::Prng rand;

   constexpr auto BitShiftAmount = 16_u64;
   static_assert(BitShiftAmount <= 32_u64, "Must contain at most uint32 # of bits");

   std::vector<uint64> u32Bins(1_u64 << BitShiftAmount, 0_u64);
   std::vector<uint64> u64Bins(1_u64 << BitShiftAmount, 0_u64);
   std::vector<uint64> f32Bins(1_u64 << BitShiftAmount, 0_u64);
   std::vector<uint64> f64Bins(1_u64 << BitShiftAmount, 0_u64);

   constexpr auto NIters = 100'000_u64;

   for (auto i = 0_u64; i < NIters; ++i)
   {
      u32Bins[rand.gen<uint32>() >> (32_u32 - BitShiftAmount)] += 1_u64;
      u64Bins[rand.gen<uint64>() >> (64_u32 - BitShiftAmount)] += 1_u64;

      f32Bins[static_cast<uint64>(
         rand.gen<float32>() * static_cast<float32>(f32Bins.size()))] += 1_u64;
      f64Bins[static_cast<uint64>(
         rand.gen<float64>() * static_cast<float64>(f64Bins.size()))] += 1_u64;
   }

   return {
      {"u32Bins", u32Bins},
      {"u64Bins", u64Bins},
      {"f32Bins", f32Bins},
      {"f64Bins", f64Bins}
   };
}

File: unittests/ym/common/rng/testsuite.py
##
# @file    testsuite.py
# @version 1.0.0
# @author  Forrest Jablonski
#

import os
import sys
import unittest

import numpy as np
from scipy.stats import binom as bn

try:
   # @note Grabs the first directory in the chain named unittests/.
   sys.path.append(os.path.join(os.getcwd().split("unittests")[0], "unittests/"))
   import testsuitebase
except:
   print("Cannot import testsuitebase - path set correctly?")
   sys.exit(1)

try:
   import cppyy
except:
   print("Cannot import cppyy - started the venv?")
   sys.exit(1)

class TestSuite(testsuitebase.TestSuiteBase):
   """
   Collection of all tests for PRNG.
   """

   @classmethod
   def setUpClass(cls):
      """
      Acting constructor.
      """

      super().setUpBaseClass(filepath="ym/common",
                             filename="rng")

   @classmethod
   def tearDownClass(cls):
      """
      Acting destructor.
      """
      pass

   def setUp(self):
      """
      Set up logic that is run before each test.
      """
      pass

   def tearDown(self):
      """
      Tear down logic that is run after each test.
      """
      pass

   def test_ZerosAndOnes(self):
      """
      Analyzes results from test case.
      """
      from cppyy.gbl import std # type:ignore
      from cppyy.gbl import ym  # type:ignore

      results = self.run_test_case("ZerosAndOnes")

      set_bit_vector = results.get[std.vector[std.pair[ym.uint64, ym.uint64]]]("SetBitVector")
      for set_bits in set_bit_vector:
         nTrials    = set_bits.first
         kSuccesses = set_bits.second
         assumedProb = 0.5
         probOfResultOrWorse = bn.cdf(kSuccesses, nTrials, assumedProb)
         print(f"For {nTrials} trials, prob of {kSuccesses} or worse {(probOfResultOrWorse*100):0.2f}%")

   def test_UniformBins(self):
      """
      Analyzes results from test case.
      """
      from cppyy.gbl import std # type:ignore
      from cppyy.gbl import ym  # type:ignore

      results = self.run_test_case("UniformBins")

      for bin_name in ["u32", "u64", "f32", "f64"]:
         bin_counts = np.array(results.get[std.vector[ym.uint64]](f"{bin_name}Bins"))
         print(f"std of {bin_name} bins {bin_counts.std()}")
         print(f"min of {bin_name} bins {bin_counts.min()}")
         print(f"max of {bin_name} bins {bin_counts.max()}")
         print(f"ave of {bin_name} bins {bin_counts.mean()}")

# kick-off
if __name__ == "__main__":
   if os.path.basename(os.getcwd()) != "rng":
      print("Needs to be run in the rng/ directory")
      sys.exit(1)

   unittest.main()
